Governance.md

```diff
@@ -0,0 +1,78 @@
+# Governance of the rules_scala
+
+These bazel rules for scala are under the bazelbuild org but governed by a
+contributors from the community.
+
+## Goals of this document
+
+Clarify:
+0. values of the project.
+1. how to get a PR merged
+2. how long you should need to wait for a review
+3. how to resolve conflicts in project direction
+4. how to make larger and disruptive changes
+
+## Values of rules_scala
+
+1. Correctness: bazel values fully reproducible builds which are reliable and hermetic (only depend on the current state of your repository)
+2. Speed: we want to build as fast as possible. In a large repository, we believe caching and remote parallel builds are the only way to achieve this. Without correctness, we can't cache and use remote builds, so correctness enables speed. We will almost never trade off correctness for speed.
+3. Usability: we want it to be as easy as possible to adopt and to maintain a build with these rules. Usability is a lower priority than correctness or speed. We will virtually never trade either correctness or speed for more usability. Usability includes stability: we seek to absolutely minimize breaking changes. We will prefer to keep some warts or deprecated methods than force users to constantly make changes to keep their builds on the latest version of the rules.
```

Comment:
I think that as the official Bazel solution for Scala, we should include wording around supporting the breadth of the Scala ecosystem. This probably falls under "Usability", but I think it's a priority that should be on-par with Correctness and Speed. 

I fear that the claim that we'll "virtually never trade either correctness or speed for more usability" will run counter to this goal, however. Maybe something along the lines of: "Should usability be at odds with speed or correctness, the latter will take precedence by default. However, we will permit usability to be achieved via optional configuration." wdyt?

Reply:
Could you elaborate on breadth, it feels unusual to me that it should have the same weighting to correctness. We should only imo support the width of the scala eco system we can do correctly and at a reasonable amount of speed. Otherwise the rules are really prone to being not really enterprise grade software since we won't have maintainers who care about lots of the features(scala-native, scala-js come to mind here) and we will start having unmaintained pieces of the repo pretty quickly. (Hi there scrooge based services)

Reply:
To be clear: this isn't some trick to go back on the agreement we made to accept a PR that adds option-in use of zinc. This is about going forward. This issue can come up again and again, and in my opinion we should be working to remove support for zinc by building migration tools and working with bazel to get first class support.

Using bazel to effectively wrap another build tool should not be our goal here. Our goal is to make the first-class bazel approach work. Not be a special scala ghetto where we do things outside of the bazel model for expedience.

Reply:
I think it's important to offer a solution that existing Scala projects can easily and cheaply adopt. Whether this is via optionally supporting existing Scala frameworks or via providing tooling that facilitates the migration doesn't matter so much, but the principle of supporting the breadth of the community stands. 

Currently, migrating to Bazel can be extremely costly. Without incremental compilation, we probably wouldn't have done so. Tooling that facilitates adoption is still far off, but in the meantime, configurability can mitigate the issue.

This isn't just about zinc--it's an issue that might come up again and block a different set of Scala users from adopting Bazel (or force them to fork/rewrite the rules).

Also, note that I'm not proposing that the maintainers implement support for the various parties of the Scala ecosystem--I just think that the ruleset should welcome off-by-default, well-tested support going forward, even if it's not as fast and correct as we'd like the principal solution to be.

Reply:
Agree that we should work towards providing an excellent Bazel + Scala experience. At the same time, I'm not sure that leveraging Bazel's configurability to make the tool more appealing to a broader audience makes the project a 'scala ghetto.' 

The more we can leverage from the Scala build community, the less we have to build an maintain. For example, the sbt test interface would make interacting with Scala testing frameworks easier. Perhaps sbt's compiler-bridge could be used as well, so we don't need to manually add support for every new compiler.

Reply:
@SrodriguezO you say: "Tooling that facilitates adoption is still far off" can you expand on that? My view is that people are pushing really hard on zinc, but not even looking into existing approaches to migration.

Some approaches:
1. https://github.com/wix/exodus
2. https://github.com/stripe/sbt-bazel
3. simplistic migration tool from maven in bazel-deps: https://github.com/johnynek/bazel-deps/blob/master/src/scala/com/github/johnynek/bazel_deps/maven/Tool.scala#L335
4. using scala-sculpt to emit a file dag: https://github.com/lightbend/scala-sculpt/pull/61

It's relatively easy to partially tool a migration (many have done it), and if we put much energy at all, I think it is totally tractable. This is why I'm negative on assuming we have to constantly compromise the bazel {fast, correct}-choose two model. We made some basic tooling, we did our migration. Wix did the same. Lucid seems very interested in this problem. Would you all be interested in following up to tool this more? I think that gives us more lift than the local maximum of incremental builds, which help you today, but make your CI very slow and hinder (by enabling giant targets) the effectiveness of caching or remote execution.

Reply:
I meant tooling that allows users to quickly and cheaply experience the benefits of Bazel. `sbt-bazel` converts an sbt project to Bazel, but this isn't enough for large projects--without 1:1:1 or incremental compilation, the developer experience in Bazel is not great. 

Automating 1:1:1 would be phenomenal, but one runs into circular dependencies and other issues and ends up having to do a lot of manual resolution in the end.

>We made some basic tooling, we did our migration. Wix did the same. 

We're glad this approach worked for you, but this shouldn't be something anyone with a different use-case should have to do before even getting a feel for Bazel… `wix/exodus` has over 900 commits.

As far as incremental builds making CI slow, we have not experienced this. But please keep in mind that the configurability/usability argument extends beyond incremental builds (that's just the unsupported use-case we encountered).

Reply:
re: breadth, I don't know how to handle it super well, but it seems the project needs to be careful about adopting things without ongoing support. We saw this in the Windows build as of recently: it was failing for some time and there's not enough support or expertise around to help fix it.

Reply:
re breadth- I agree with Ian and Long. If people with commitment come then we should definitely be welcoming but I'm really not sure it's a project goal. We have a lot of work cut out for us in the current challenges.
Re the general discussion- I think we need to add an accessibility value which is rather low which talks about easing the bar by Bazel mechanisms while accepting current compromises.
Re `wix/exodus has over 900 commits`- we've exported full history to be open. More so this just means we've solved a lot of the problems already (1:1:1 with cycles support).

Reply:
Usability: We want it to be as easy as possible to adopt and to maintain a build with these rules. However, in the pursuit of usability, we cannot sacrifice correctness or speed. We will seek to implement the best tools and methods and avoid shortcuts when creating first class support in Bazel for Scala.
Stability: We seek to minimize breaking changes to those that are absolutely necessary. We will avoid forcing users to constantly make changes to keep their builds on the latest version of the rules.